executor/insert_common.go

```diff
@@ -778,18 +792,36 @@ func insertRowsFromSelectWorker(ctx context.Context, base insertCommon, batchSiz
 		case <-ctx.Done():
 			return nil
 		}
-		rows = make([][]types.Datum, 0, chk.NumRows())
+		rows = rows[:0]
+		rowsCacheIndex = 0
 		iter := chunk.NewIterator4Chunk(chk)
 		for innerChunkRow := iter.Begin(); innerChunkRow != iter.End(); innerChunkRow = iter.Next() {
-			innerRow := innerChunkRow.GetDatumRow(fields)
+			length := innerChunkRow.Len()
+			if len(datumRow) < length {
+				diff := length - len(datumRow)
+				for i := 0; i < diff; i++ {
+					datumRow = append(datumRow, types.Datum{}) //nolint:makezero
+				}
+			} else {
+				datumRow = datumRow[:length]
+				for i := 0; i < length; i++ {
+					datumRow[i] = types.Datum{}
+				}
+			}
+			datumRow := innerChunkRow.GetDatumRowWithBuffer(fields, datumRow)
 			var row []types.Datum
-			row, err = e.getRow(ctx, innerRow)
+			if rowsCacheIndex >= len(rowsCache) {
+				rowsCache = append(rowsCache, make([]types.Datum, columnLen)) //nolint:makezero
+			}
+			row = rowsCache[rowsCacheIndex]
+			rowsCacheIndex++
+			row, err = e.getRowWithBuffer(ctx, datumRow, row, hasValue)
```

Comment:
The generated row may use the byte reference in the `Datum` object which is easy to make mistakes, by now seems the cached `Datum` object is cleared each time so it's safe?